plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java

```diff
@@ -54,21 +72,16 @@
    * operation and display a general message that login is required but was cancelled or failed.
    *
    * Must be called from a UI context.
-   *
-   * @param shellProvider provides a shell for the login screen if login is necessary
-   * @throws IOException can be thrown by the underlying Login API request (e.g., network
-   *     error from the transport layer while sending/receiving a HTTP request/response.)
    */
-  public Credential getActiveCredential(IShellProvider shellProvider) throws IOException {
-    Credential credential = getCachedActiveCredential();
-
-    if (credential == null) {
-      credential = logIn(shellProvider);
-
-      IEclipseContext eclipseContext = PlatformUI.getWorkbench().getService(IEclipseContext.class);
-      eclipseContext.set(STASH_OAUTH_CRED_KEY, credential);
+  public Credential getActiveCredential() {
+    // TODO: holding a lock for a long period of time (especially when waiting for UI events)
```

Comment:
That's why I was thinking about locks. If you use the flag, and it's false (meaning no login is in progress), by the time you get to the next line and start the synchronized block, it can be changed to true by a concurrent thread. Possibly a very small chance, but it's still there.